src/permission-set.js

```diff
@@ -16,523 +16,809 @@
+   * @param webId {String} URL of an agent for which this permission applies
+   * @param accessMode {String|Array<String>} One or more access modes
+   * @param [origin] {String|Array<String>} One or more allowed origins (optional)
+   * @return {PermissionSet} Returns self (chainable)
    */
-  this.origin = options.origin
+  addPermission (webId, accessMode, origin) {
+    if (!webId) {
+      throw new Error('addPermission() requires a valid webId')
+    }
+    if (!accessMode) {
+      throw new Error('addPermission() requires a valid accessMode')
+    }
+    var auth = new Authorization(this.resourceUrl, this.isAuthInherited())
+    auth.setAgent(webId)
+    auth.addMode(accessMode)
+    if (origin) {
+      auth.addOrigin(origin)
+    }
+    this.addAuthorization(auth)
+    return this
+  }
+
   /**
-   * Solid REST client (optionally injected), used by save() and clear().
-   * @type {SolidWebClient}
+   * Adds a given authorization to the "lookup by agent id" index.
+   * Enables lookups via `findAuthByAgent()`.
+   * @method addToAgentIndex
+   * @private
+   * @param webId {String} Agent's Web ID
+   * @param accessType {String} Either `accessTo` or `default`
+   * @param resourceUrl {String}
+   * @param authorization {Authorization}
    */
-  this.webClient = options.webClient
-  // Optionally initialize from a given parsed graph
-  if (options.graph) {
-    this.initFromGraph(options.graph)
+  addToAgentIndex (webId, accessType, resourceUrl, authorization) {
+    let agents = this.index.agents
+    if (!agents[webId]) {
+      agents[webId] = {}
+    }
+    if (!agents[webId][accessType]) {
+      agents[webId][accessType] = {}
+    }
+    if (!agents[webId][accessType][resourceUrl]) {
+      agents[webId][accessType][resourceUrl] = authorization
+    } else {
+      agents[webId][accessType][resourceUrl].mergeWith(authorization)
+    }
   }
-}
 
-/**
- * Adds a given Authorization instance to the permission set.
- * Low-level function, clients should use `addPermission()` instead, in most
- * cases.
- * @method addAuthorization
- * @param auth {Authorization}
- * @return {PermissionSet} Returns self (chainable)
- */
-function addAuthorization (auth) {
-  var hashFragment = auth.hashFragment()
-  if (hashFragment in this.authorizations) {
-    // An authorization for this agent and resource combination already exists
-    // Merge the incoming access modes with its existing ones
-    this.authorizations[hashFragment].mergeWith(auth)
-  } else {
-    this.authorizations[hashFragment] = auth
+  /**
+   * Adds a given authorization to the "lookup by group id" index.
+   * Enables lookups via `findAuthByAgent()`.
+   * @method addToGroupIndex
+   * @private
+   * @param webId {String} Group's Web ID
+   * @param accessType {String} Either `accessTo` or `default`
+   * @param resourceUrl {String}
+   * @param authorization {Authorization}
+   */
+  addToGroupIndex (webId, accessType, resourceUrl, authorization) {
```

Comment:
Sure, if you feel strongly about it. My initial instinct was to go this route (with two similar methods) so that a text search for `groups` or `agents` would find these methods.